File: cricrelay-mobile/android/feature/umpire-scorer/src/main/java/uk/co/cricrelay/mobile/feature/umpire/UmpireScorerViewModel.kt
package uk.co.cricrelay.mobile.feature.umpire

import android.content.Context
import androidx.lifecycle.ViewModel
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import org.json.JSONArray
import org.json.JSONObject
import uk.co.cricrelay.shared.model.UmpireBallEvent
import uk.co.cricrelay.shared.model.UmpireDeliveryMode
import uk.co.cricrelay.shared.model.UmpireScorerState
import javax.inject.Inject

@HiltViewModel
class UmpireScorerViewModel @Inject constructor(
    @ApplicationContext private val context: Context,
) : ViewModel() {

    private val _state = MutableStateFlow(UmpireScorerState())
    val uiState: StateFlow<UmpireScorerState> = _state.asStateFlow()

    private val _newOverEvent = MutableSharedFlow<Unit>(extraBufferCapacity = 1)
    val newOverEvent: SharedFlow<Unit> = _newOverEvent.asSharedFlow()

    private val history = ArrayDeque<UmpireScorerState>()

    init {
        loadState()
    }

    fun onRunsPressed(runs: Int) {
        pushHistory()
        val s = _state.value
        val isLegal = s.deliveryMode == UmpireDeliveryMode.NORMAL ||
                s.deliveryMode == UmpireDeliveryMode.BYE ||
                s.deliveryMode == UmpireDeliveryMode.LEGBYE
        val runsScored = if (s.deliveryMode == UmpireDeliveryMode.WIDE ||
            s.deliveryMode == UmpireDeliveryMode.NOBALL) runs + 1 else runs
        val ball = UmpireBallEvent(
            label = buildLabel(s.deliveryMode, s.pendingWicket, runs),
            totalRuns = runsScored,
            isLegal = isLegal,
            isWicket = s.pendingWicket,
        )
        val newLegal = s.legalBallCount + if (isLegal) 1 else 0
        _state.value = s.copy(
            totalRuns = s.totalRuns + runsScored,
            totalWickets = s.totalWickets + if (s.pendingWicket) 1 else 0,
            currentOverBalls = s.currentOverBalls + ball,
            legalBallCount = newLegal,
            deliveryMode = UmpireDeliveryMode.NORMAL,
            pendingWicket = false,
        )
        saveState()
        if (newLegal >= 6) _newOverEvent.tryEmit(Unit)
    }

    fun onWidePressed() {
        val s = _state.value
        _state.value = s.copy(
            deliveryMode = if (s.deliveryMode == UmpireDeliveryMode.WIDE) UmpireDeliveryMode.NORMAL else UmpireDeliveryMode.WIDE,
        )
    }

    fun onNoBallPressed() {
        val s = _state.value
        _state.value = s.copy(
            deliveryMode = if (s.deliveryMode == UmpireDeliveryMode.NOBALL) UmpireDeliveryMode.NORMAL else UmpireDeliveryMode.NOBALL,
        )
    }

    fun onByePressed() {
        val s = _state.value
        _state.value = s.copy(
            deliveryMode = if (s.deliveryMode == UmpireDeliveryMode.BYE) UmpireDeliveryMode.NORMAL else UmpireDeliveryMode.BYE,
        )
    }

    fun onLegByePressed() {
        val s = _state.value
        _state.value = s.copy(
            deliveryMode = if (s.deliveryMode == UmpireDeliveryMode.LEGBYE) UmpireDeliveryMode.NORMAL else UmpireDeliveryMode.LEGBYE,
        )
    }

    fun onWicketToggle() {
        val s = _state.value
        _state.value = s.copy(pendingWicket = !s.pendingWicket)
    }

    fun onUndo() {
        val prev = history.removeLastOrNull() ?: return
        _state.value = prev
        saveState()
    }

    fun onEndOver() {
        pushHistory()
        val s = _state.value
        _state.value = s.copy(
            completedOvers = s.completedOvers + 1,
            currentOverBalls = emptyList(),
            legalBallCount = 0,
            deliveryMode = UmpireDeliveryMode.NORMAL,
            pendingWicket = false,
        )
        saveState()
    }

    fun onReset() {
        history.clear()
        _state.value = UmpireScorerState()
        saveState()
    }

    private fun pushHistory() {
        history.addLast(_state.value)
        if (history.size > 50) history.removeFirst()
    }

    private fun buildLabel(mode: UmpireDeliveryMode, wicket: Boolean, runs: Int): String {
        val base = when (mode) {
            UmpireDeliveryMode.NORMAL -> if (runs == 0) "·" else "$runs"
            UmpireDeliveryMode.WIDE -> if (runs == 0) "Wd" else "Wd+$runs"
            UmpireDeliveryMode.NOBALL -> if (runs == 0) "Nb" else "Nb+$runs"
            UmpireDeliveryMode.BYE -> if (runs == 0) "B" else "B+$runs"
            UmpireDeliveryMode.LEGBYE -> if (runs == 0) "Lb" else "Lb+$runs"
        }
        return if (wicket) "$base+W" else base
    }

    // ── Persistence ───────────────────────────────────────────────────────────

    private fun saveState() {
        val s = _state.value
        val prefs = context.getSharedPreferences("umpire_scorer", Context.MODE_PRIVATE)
        val ballsArray = JSONArray().apply {
            s.currentOverBalls.forEach { ball ->
                put(JSONObject().apply {
                    put("label", ball.label)
                    put("totalRuns", ball.totalRuns)
                    put("isLegal", ball.isLegal)
                    put("isWicket", ball.isWicket)
                })
            }
        }
        prefs.edit().apply {
            putInt("totalRuns", s.totalRuns)
            putInt("totalWickets", s.totalWickets)
            putInt("completedOvers", s.completedOvers)
            putInt("legalBallCount", s.legalBallCount)
            putString("currentOverBalls", ballsArray.toString())
        }.apply()
    }

    private fun loadState() {
        val prefs = context.getSharedPreferences("umpire_scorer", Context.MODE_PRIVATE)
        if (!prefs.contains("totalRuns")) return
        val balls = mutableListOf<UmpireBallEvent>()
        runCatching {
            val arr = JSONArray(prefs.getString("currentOverBalls", "[]") ?: "[]")
            for (i in 0 until arr.length()) {
                val obj = arr.getJSONObject(i)
                balls.add(
                    UmpireBallEvent(
                        label = obj.getString("label"),
                        totalRuns = obj.getInt("totalRuns"),
                        isLegal = obj.getBoolean("isLegal"),
                        isWicket = obj.getBoolean("isWicket"),
                    ),
                )
            }
        }
        _state.value = UmpireScorerState(
            totalRuns = prefs.getInt("totalRuns", 0),
            totalWickets = prefs.getInt("totalWickets", 0),
            completedOvers = prefs.getInt("completedOvers", 0),
            legalBallCount = prefs.getInt("legalBallCount", 0),
            currentOverBalls = balls,
        )
    }
}

File: cricrelay-mobile/ios/Features/UmpireScorer/UmpireScorerViewModel.swift
import Foundation
import Combine

enum UmpireDeliveryMode: String, Codable {
    case normal, wide, noBall, bye, legBye
}

struct UmpireBallEvent: Codable, Identifiable {
    let id: UUID
    let label: String
    let totalRuns: Int
    let isLegal: Bool
    let isWicket: Bool

    init(label: String, totalRuns: Int, isLegal: Bool, isWicket: Bool) {
        self.id = UUID()
        self.label = label
        self.totalRuns = totalRuns
        self.isLegal = isLegal
        self.isWicket = isWicket
    }
}

struct UmpireScorerPersistable: Codable {
    var totalRuns: Int
    var totalWickets: Int
    var completedOvers: Int
    var legalBallCount: Int
    var currentOverBalls: [UmpireBallEvent]
}

@MainActor
final class UmpireScorerViewModel: ObservableObject {
    @Published var totalRuns = 0
    @Published var totalWickets = 0
    @Published var completedOvers = 0
    @Published var currentOverBalls: [UmpireBallEvent] = []
    @Published var legalBallCount = 0
    @Published var deliveryMode: UmpireDeliveryMode = .normal
    @Published var pendingWicket = false

    private var history: [[UmpireBallEvent]] = []
    private let persistKey = "umpire_scorer_state"

    init() {
        load()
    }

    func onRunsPressed(_ runs: Int) {
        history.append(currentOverBalls)
        let isLegal = deliveryMode == .normal || deliveryMode == .bye || deliveryMode == .legBye
        let runsScored = (deliveryMode == .wide || deliveryMode == .noBall) ? runs + 1 : runs
        let ball = UmpireBallEvent(
            label: buildLabel(mode: deliveryMode, wicket: pendingWicket, runs: runs),
            totalRuns: runsScored,
            isLegal: isLegal,
            isWicket: pendingWicket
        )
        currentOverBalls.append(ball)
        totalRuns += runsScored
        if pendingWicket { totalWickets += 1 }
        if isLegal { legalBallCount += 1 }
        deliveryMode = .normal
        pendingWicket = false
        save()
    }

    func onWidePressed() {
        deliveryMode = deliveryMode == .wide ? .normal : .wide
    }

    func onNoBallPressed() {
        deliveryMode = deliveryMode == .noBall ? .normal : .noBall
    }

    func onByePressed() {
        deliveryMode = deliveryMode == .bye ? .normal : .bye
    }

    func onLegByePressed() {
        deliveryMode = deliveryMode == .legBye ? .normal : .legBye
    }

    func onWicketToggle() {
        pendingWicket.toggle()
    }

    func onUndo() {
        guard let prev = history.popLast() else { return }
        let undone = currentOverBalls.last
        currentOverBalls = prev
        if let ball = undone {
            totalRuns -= ball.totalRuns
            if ball.isWicket { totalWickets -= 1 }
            if ball.isLegal { legalBallCount -= 1 }
        }
        deliveryMode = .normal
        pendingWicket = false
        save()
    }

    func onEndOver() {
        completedOvers += 1
        currentOverBalls = []
        legalBallCount = 0
        deliveryMode = .normal
        pendingWicket = false
        history.removeAll()
        save()
    }

    func onReset() {
        totalRuns = 0
        totalWickets = 0
        completedOvers = 0
        currentOverBalls = []
        legalBallCount = 0
        deliveryMode = .normal
        pendingWicket = false
        history.removeAll()
        save()
    }

    var modeIndicatorText: String {
        switch (deliveryMode, pendingWicket) {
        case (.normal, false): return "Tap runs — or pick type first"
        case (.normal, true): return "WICKET pending — tap runs"
        case (let m, true): return "\(m.rawValue.uppercased()) + WICKET — tap runs"
        case (let m, false): return "\(m.rawValue.uppercased()) — tap runs"
        }
    }

    // MARK: - Helpers

    private func buildLabel(mode: UmpireDeliveryMode, wicket: Bool, runs: Int) -> String {
        let base: String
        switch mode {
        case .normal:  base = runs == 0 ? "·" : "\(runs)"
        case .wide:    base = runs == 0 ? "Wd" : "Wd+\(runs)"
        case .noBall:  base = runs == 0 ? "Nb" : "Nb+\(runs)"
        case .bye:     base = runs == 0 ? "B" : "B+\(runs)"
        case .legBye:  base = runs == 0 ? "Lb" : "Lb+\(runs)"
        }
        return wicket ? "\(base)+W" : base
    }

    // MARK: - Persistence

    private func save() {
        let data = UmpireScorerPersistable(
            totalRuns: totalRuns,
            totalWickets: totalWickets,
            completedOvers: completedOvers,
            legalBallCount: legalBallCount,
            currentOverBalls: currentOverBalls
        )
        if let encoded = try? JSONEncoder().encode(data) {
            UserDefaults.standard.set(encoded, forKey: persistKey)
        }
    }

    private func load() {
        guard let data = UserDefaults.standard.data(forKey: persistKey),
              let saved = try? JSONDecoder().decode(UmpireScorerPersistable.self, from: data)
        else { return }
        totalRuns = saved.totalRuns
        totalWickets = saved.totalWickets
        completedOvers = saved.completedOvers
        legalBallCount = saved.legalBallCount
        currentOverBalls = saved.currentOverBalls
    }
}
